ee/desktop/menu/menu.go

```diff
@@ -1,36 +1,174 @@
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
 	"github.com/kolide/launcher/ee/desktop/assets"
 )
 
-func Init(hostname string) {
-	onReady := func() {
-		systray.SetTemplateIcon(assets.KolideDesktopIcon, assets.KolideDesktopIcon)
-		systray.SetTooltip("Kolide")
+// MenuData encapsulates a systray icon and accessible menu items
+type MenuData struct {
+	Tooltip string         `json:"tooltip,omitempty"`
+	Items   []MenuItemData `json:"items"`
+}
 
-		versionItem := systray.AddMenuItem(fmt.Sprintf("Version %s", version.Version().Version), "")
-		versionItem.Disable()
+// MenuItemData represents a menu item, optionally containing sub menu items
+type MenuItemData struct {
+	Label       string         `json:"label,omitempty"`
+	Tooltip     string         `json:"tooltip,omitempty"`
+	Disabled    bool           `json:"disabled,omitempty"`
+	NonProdOnly bool           `json:"nonProdOnly,omitempty"`
+	IsSeparator bool           `json:"isSeparator,omitempty"`
+	Action      MenuItemAction `json:"action,omitempty"`
+	Items       []MenuItemData `json:"items,omitempty"`
+}
 
-		// if prod environment, return
-		if hostname == "k2device-preprod.kolide.com" || hostname == "k2device.kolide.com" {
-			return
-		}
+// MenuItemAction encapsulates what action should be performed when a menu item is invoked
+type MenuItemAction struct {
+	URL string `json:"url,omitempty"`
+}
+
+// MenuBuilder is an interface a menu parser can use to specify how the systray menu is built
+type MenuBuilder interface {
+	SetTooltip(tooltip string)
+	AddMenuItem(label, tooltip string, disabled, nonProdOnly bool, action MenuItemAction, parent any) any
+	AddSeparator()
+}
+
+type menu struct {
+	logger    log.Logger
+	hostname  string
+	doneChans []chan bool
+}
+
+func New(logger log.Logger, hostname string) *menu {
+	m := &menu{
+		logger:   logger,
+		hostname: hostname,
+	}
 
-		// in non prod environment
+	return m
+}
+
+func (m *menu) Init(buildMenu func()) {
+	// buildMenu will be invoked after the systray has been initialized
+	// Before the systray exits, cleanup the goroutines
+	systray.Run(buildMenu, m.cleanup)
+}
+
+func (m *menu) Build(menu *MenuData) {
+	// Remove all menu items each time we rebuild the menu
+	systray.ResetMenu()
+
+	// Even though the menu items have been removed, we still have goroutines hanging around
+	m.cleanup()
+
+	parseMenuData(menu, m)
+}
+
+func (m *menu) SetTooltip(tooltip string) {
+	systray.SetTooltip(tooltip)
+
+	if m.isProd() {
+		systray.SetTemplateIcon(assets.KolideDesktopIcon, assets.KolideDesktopIcon)
+	} else {
 		systray.SetTemplateIcon(assets.KolideDebugDesktopIcon, assets.KolideDebugDesktopIcon)
-		systray.AddSeparator()
-		systray.AddMenuItem("--- DEBUG ---", "").Disable()
-		systray.AddMenuItem(fmt.Sprintf("Hostname: %s", hostname), "").Disable()
 	}
+}
 
-	systray.Run(onReady, func() {})
+func (m *menu) AddMenuItem(label, tooltip string, disabled, nonProdOnly bool, action MenuItemAction, parent any) any {
```

Comment:
I'm confused about what calls this, and why it has the signature it has

Reply:
This could definitely use some better docs and function comments.

My intention was to use `MenuBuilder` interface as an abstraction over the underlying systray implementation. Ideally the menu data schema can be parsed independently of whatever underlying framework we use to actually build the menu. The parsing functions use the `MenuBuilder` functions like this one to instruct how the menu should be built without having a dependency on the fyne systray. If we ever swap out systray for say, the v2 one, my hope is that only the menu builder code is changed, and the parsing stays agnostic to the underlying menu dependency.

A non-obvious part of this signature is that in order to support sub menus, we need a reference to the parent menu so we can add menu items to it. This function always returns an `any` type reference to the menu item it created - the parsing code uses this when constructing sub menus by passing it back to the builder, and that's how the builder knows to add the menu item as a child menu item of that parent, instead of the top-level menu item (which doesn't require any menu item reference).